// @parcel/resolver-glob/lib/GlobResolver.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _plugin() {
  const data = require("@parcel/plugin");

  _plugin = function () {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("@parcel/utils");

  _utils = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _diagnostic() {
  const data = _interopRequireDefault(require("@parcel/diagnostic"));

  _diagnostic = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = new (_plugin().Resolver)({
  async resolve({
    dependency,
    options,
    specifier,
    pipeline
  }) {
    var _dependency$resolveFr, _dependency$meta;

    if (!(0, _utils().isGlob)(specifier)) {
      return;
    }

    let sourceAssetType = (0, _nullthrows().default)(dependency.sourceAssetType);
    let sourceFile = (0, _nullthrows().default)((_dependency$resolveFr = dependency.resolveFrom) !== null && _dependency$resolveFr !== void 0 ? _dependency$resolveFr : dependency.sourcePath);
    let error;

    if (sourceAssetType !== 'js' && sourceAssetType !== 'css') {
      error = `Glob imports are not supported in ${sourceAssetType} files.`;
    } else if (dependency.specifierType === 'url' && !((_dependency$meta = dependency.meta) !== null && _dependency$meta !== void 0 && _dependency$meta.isCSSImport)) {
      error = 'Glob imports are not supported in URL dependencies.';
    }

    if (error) {
      throw new (_diagnostic().default)({
        diagnostic: {
          message: error,
          codeFrames: dependency.loc ? [{
            codeHighlights: [{
              start: dependency.loc.start,
              end: dependency.loc.end
            }]
          }] : undefined
        }
      });
    }

    specifier = _path().default.resolve(_path().default.dirname(sourceFile), specifier);
    let normalized = (0, _utils().normalizeSeparators)(specifier);
    let files = await (0, _utils().glob)(normalized, options.inputFS, {
      onlyFiles: true
    });

    let dir = _path().default.dirname(specifier);

    let results = files.map(file => {
      let relative = (0, _utils().relativePath)(dir, file);

      if (pipeline) {
        relative = `${pipeline}:${relative}`;
      }

      return [file, relative];
    });
    let code = '';

    if (sourceAssetType === 'js') {
      let re = (0, _utils().globToRegex)(normalized, {
        capture: true
      });
      let matches = {};

      for (let [file, relative] of results) {
        let match = file.match(re);
        if (!match) continue;
        let parts = match.slice(1).filter(Boolean).reduce((a, p) => a.concat(p.split('/')), []);
        set(matches, parts, relative);
      }

      let {
        value,
        imports
      } = generate(matches, dependency.priority === 'lazy');
      code = imports + 'module.exports = ' + value;
    } else if (sourceAssetType === 'css') {
      for (let [, relative] of results) {
        code += `@import "${relative}";\n`;
      }
    }

    return {
      filePath: _path().default.join(dir, _path().default.basename(specifier, _path().default.extname(specifier)) + '.' + sourceAssetType),
      code,
      invalidateOnFileCreate: [{
        glob: normalized
      }],
      pipeline: null,
      priority: 'sync'
    };
  }

});

exports.default = _default;

function set(obj, path, value) {
  for (let i = 0; i < path.length - 1; i++) {
    let part = path[i];

    if (obj[part] == null) {
      obj[part] = {};
    }

    obj = obj[part];
  }

  obj[path[path.length - 1]] = value;
}

function generate(matches, isAsync, indent = '', count = 0) {
  if (typeof matches === 'string') {
    if (isAsync) {
      return {
        imports: '',
        value: `() => import(${JSON.stringify(matches)})`,
        count
      };
    }

    let key = `_temp${count++}`;
    return {
      imports: `const ${key} = require(${JSON.stringify(matches)});`,
      value: key,
      count
    };
  }

  let imports = '';
  let res = indent + '{';
  let first = true;

  for (let key in matches) {
    if (!first) {
      res += ',';
    }

    let {
      imports: i,
      value,
      count: c
    } = generate(matches[key], isAsync, indent + '  ', count);
    imports += `${i}\n`;
    count = c;
    res += `\n${indent}  ${JSON.stringify(key)}: ${value}`;
    first = false;
  }

  res += '\n' + indent + '}';
  return {
    imports,
    value: res,
    count
  };
}
